pkg/proc/breakpoints.go

```diff
@@ -163,36 +172,76 @@ type returnBreakpointInfo struct {
 // CheckCondition evaluates bp's condition on thread.
 func (bp *Breakpoint) CheckCondition(thread Thread) BreakpointState {
 	bpstate := BreakpointState{Breakpoint: bp, Active: false, Internal: false, CondError: nil}
-	if bp.Cond == nil && bp.internalCond == nil {
+	bpstate.checkCond(thread)
+	// Update the breakpoint hit counts.
+	if bpstate.Breakpoint != nil && bpstate.Active {
+		if g, err := GetG(thread); err == nil {
+			bpstate.HitCount[g.ID]++
+		}
+		bpstate.TotalHitCount++
+	}
+	bpstate.checkHitCond(thread)
+	return bpstate
+}
+
+func (bpstate *BreakpointState) checkCond(thread Thread) {
+	if bpstate.Cond == nil && bpstate.internalCond == nil {
 		bpstate.Active = true
-		bpstate.Internal = bp.IsInternal()
-		return bpstate
+		bpstate.Internal = bpstate.IsInternal()
+		return
 	}
 	nextDeferOk := true
-	if bp.Kind&NextDeferBreakpoint != 0 {
+	if bpstate.Kind&NextDeferBreakpoint != 0 {
 		var err error
 		frames, err := ThreadStacktrace(thread, 2)
 		if err == nil {
 			nextDeferOk = isPanicCall(frames)
 			if !nextDeferOk {
-				nextDeferOk, _ = isDeferReturnCall(frames, bp.DeferReturns)
+				nextDeferOk, _ = isDeferReturnCall(frames, bpstate.DeferReturns)
 			}
 		}
 	}
-	if bp.IsInternal() {
+	if bpstate.IsInternal() {
 		// Check internalCondition if this is also an internal breakpoint
-		bpstate.Active, bpstate.CondError = evalBreakpointCondition(thread, bp.internalCond)
+		bpstate.Active, bpstate.CondError = evalBreakpointCondition(thread, bpstate.internalCond)
 		bpstate.Active = bpstate.Active && nextDeferOk
 		if bpstate.Active || bpstate.CondError != nil {
 			bpstate.Internal = true
-			return bpstate
+			return
 		}
 	}
-	if bp.IsUser() {
+	if bpstate.IsUser() {
 		// Check normal condition if this is also a user breakpoint
-		bpstate.Active, bpstate.CondError = evalBreakpointCondition(thread, bp.Cond)
+		bpstate.Active, bpstate.CondError = evalBreakpointCondition(thread, bpstate.Cond)
 	}
-	return bpstate
+}
+
+// checkHitCond evaluates bp's hit condition on thread.
+func (bpstate *BreakpointState) checkHitCond(thread Thread) {
+	if bpstate.HitCond == nil || !bpstate.Active || bpstate.Internal {
+		return
+	}
+	bpstate.setHitCondTotalHitCount()
+	// Evaluate the breakpoint condition and set CondError if it has not
+	// already been set.
+	var err error
+	bpstate.Active, err = evalBreakpointCondition(thread, bpstate.HitCond)
+	if bpstate.CondError == nil {
+		bpstate.CondError = err
+	}
+}
+
+func (bpstate *BreakpointState) setHitCondTotalHitCount() {
```

Comment:
Is this to support things like `(hitcount >= 9) && (hitcount <= 15)`? It won't work because it will only replace the leftmost variable.

If this really needs arbitrary expressions involving the hitcount it should be a fake variable, like runtime.curg (see the ast.SelectorExpr in evalAST).

If it just needs to compare the hit count with an integer then maybe it doesn't need to be a BinaryExpr at all, just an operator ("<", ">=", etc.) and an integer argument. And it gets evaluated in a `switch` statement, without involving an EvalScope.

Reply:
This was to support things like (`hitcount % 100 == 0`), though this could also be supported as `OP NUMBER`

Reply:
I'll update this logic to accept `NUMBER` or `OP NUMBER`! I think that makes more sense for now. 

If there is a use case to support other expressions we can change this at a later point after getting user feedback.